tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,159 @@
+    my $downloaded_xml_dir = "/tmp/download_vm_xml";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        unless (script_run("ls $downloaded_xml_dir/$guest.xml") == 0) {
+            assert_script_run "mkdir -p $downloaded_xml_dir";
+            assert_script_run "virsh dumpxml --inactive $guest > $downloaded_xml_dir/$guest.xml";
+        }
+    }
+
+    foreach my $guest (keys %virt_autotest::common::guests) {
+
+        #irqbalance test run on sle12sp3+ guest
+        return if guest_is_sle($guest, '<=12-sp2');
+
+        record_info("Test $guest");
+        prepare_guest_for_irqbalance($guest);
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        #Check if network interrupts are distributed over CPUs in guest initally as irqbalance is enabled by default
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            die("The value of network interrupts for CPU" . $cpu_id . " is 0") if ($initial_total_irqs_on_cpu[$cpu_id] == 0);
+        }
+
+        # Bind all NIC IRQs to cpu0 in guest
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        my $cat_affinity_files_cmdline = "ssh root\@$guest \"cd /proc/irq && cat";
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+            $cat_affinity_files_cmdline .= " $irq_id/smp_affinity";
+        }
+        $cat_affinity_files_cmdline .= "\"";
+        # with irqbalance.service is enabled, SMP affinity will be auto balanced even if IRQs are bound to specified CPUs in SLE15SP4
+        # while in prior releases, such as SLE15SP3 or SLE12SP5, IRQs can be bound successfully until another irqbalance is run again
+        assert_script_run("ssh root\@$guest \"irqbalance --oneshot\"") if guest_is_sle($guest, '<=15-sp3');
+        my @affinity_after_binding = split('\n', script_output($cat_affinity_files_cmdline, 60));
+        foreach my $affinity (@affinity_after_binding) {
+            if ($affinity eq 1) {
+                # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+                # So it is correct if some of them may happen to run on cpu0
+                # but it would not be correct if all IRQs were bound to cpu0 successfully.
+                if (guest_is_sle($guest, '<=15-sp3')) {
+                    record_soft_failure("One of the NIC IRQs was bound to cpu0, which may be not expected.");
+                }
+                # for SLE15SP4 (or newer release), IRQs are specified on all CPU cores
+                # fg. affinity value is '0xf' for machine with 4 CPU cores
+                else {
+                    die "It is not correct that one of the NIC IRQs was bound to cpu0 successfully with irqbalance enabled.";
+                }
```

Comment:
Sorry, I am confused here. What's the logic here and what are the expected results?

Reply:
fg. the affinities with 4 cpus with irqbalance may be(take 4 NIC IRQs as an example):
before binding: 
```
'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3' for 15sp4(https://openqa.nue.suse.com/tests/8009060#step/xen_guest_irqbalance/136); 
'cpu1', 'cpu3', 'cpu0', 'cpu2' for 15sp3(https://openqa.nue.suse.com/tests/8009059#step/xen_guest_irqbalance/282); 
'cpu8', 'cpu2', 'cpu1', 'cpu3' for 12sp5(https://openqa.nue.suse.com/tests/8009061#step/xen_guest_irqbalance/288); 
```

after binding to cpu0 WITHOUT 'irqbalance --oneshot' run:
```
'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3', 'cpu0 cpu1 cpu2 cpu3' for 15sp4; 
'cpu0', 'cpu0', 'cpu0', 'cpu0' for 15sp3; 
'cpu0', 'cpu0', 'cpu0', 'cpu0' for 12sp5
```

after binding to cpu0 WITH 'irqbalance --oneshot' run:
```
the same with the values before binding